File: 03-LSP/python/refactored.py
class Bird:
    def eat(self):
        print("Eating")


class FlyingBird(Bird):
    def fly(self):
        print("Flying")


class Sparrow(FlyingBird):
    def fly(self):
        print("Sparrow flying")


class Penguin(Bird):
    def eat(self):
        print("Penguin eating")


def make_bird_eat(bird):
    bird.eat()


def make_bird_fly(bird):
    bird.fly()


if __name__ == "__main__":
    sparrow = Sparrow()
    penguin = Penguin()

    make_bird_eat(sparrow)
    make_bird_eat(penguin)

    make_bird_fly(sparrow)

'''
Output:
Eating
Penguin eating
Sparrow flying
'''

File: 03-LSP/cpp/refactored.cpp
#include <iostream>

class Bird {
public:
    virtual void eat() = 0;
    virtual ~Bird() = default;
};

class FlyingBird : public Bird {
public:
    void eat() override {
        std::cout << "Eating\n";
    }

    virtual void fly() = 0;
};

class Sparrow : public FlyingBird {
public:
    void fly() override {
        std::cout << "Sparrow flying\n";
    }
};

class Penguin : public Bird {
public:
    void eat() override {
        std::cout << "Penguin eating\n";
    }
};

void makeBirdEat(Bird& bird) {
    bird.eat();
}

void makeBirdFly(FlyingBird& bird) {
    bird.fly();
}

int main() {
    Sparrow sparrow;
    Penguin penguin;

    makeBirdEat(sparrow);
    makeBirdEat(penguin);

    makeBirdFly(sparrow);

    return 0;
}